File: chain/TransactionBuilder.ts
import { CustomTransactionRequest } from "../core/BaseTypes/TransactionRequest.js";
import { Address } from "../core/BaseTypes/Address.js";
import WalletManager from "../core/WalletManager.js";
import ChainClient from "./ChainClient.js";
import TokenAmount from "../core/BaseTypes/TokenAmount.js";
import { CustomTransactionReceipt } from "../core/BaseTypes/TransactionReceipt.js";

export class TransactionBuilder {
  /*
    Fluent builder for transactions.

    Usage:
        const tx = (TransactionBuilder(client, wallet)
            .to(recipient)
            .value(TokenAmount.fromHuman("0.1", 18))
            .data(calldata)
            .withGasEstimate()
            .withGasPrice("high")
            .build())
    */

  static readonly transactionGasLimit = 21_000;

  private readonly client: ChainClient;
  private readonly wallet: WalletManager;
  private params = {} as {
    to: Address;
    value: TokenAmount;
    data?: Uint8Array;
    nonce?: number;
    gasLimit?: number;
    maxFeePerGas?: bigint;
    maxPriorityFee?: bigint;
    chainId?: number;
  };
  constructor(client: ChainClient, wallet: WalletManager) {
    this.client = client;
    this.wallet = wallet;
    this.reset();
  }
  reset() {
    this.params = {} as {
      to: Address;
      value: TokenAmount;
      data?: Uint8Array;
      nonce?: number;
      gasLimit?: number;
      maxFeePerGas?: bigint;
      maxPriorityFee?: bigint;
      chainId?: number;
    };
  }
  to(address: Address): TransactionBuilder {
    this.params.to = address;
    return this;
  }

  value(amount: TokenAmount): TransactionBuilder {
    this.params.value = amount;
    return this;
  }

  data(calldata: Uint8Array): TransactionBuilder {
    this.params.data = calldata;
    return this;
  }
  nonce(nonce: number): TransactionBuilder {
    this.params.nonce = nonce;
    return this;
  }
  gasLimit(limit: number): TransactionBuilder {
    this.params.gasLimit = limit;
    return this;
  }
  async withGasEstimate(buffer: number = 1.2): Promise<this> {
    const tx = new CustomTransactionRequest(this.params as any);

    const estimated = await this.client.estimateGas(tx);

    this.params.gasLimit = Math.ceil(Number(estimated) * buffer);

    return this;
  }

  async withGasPrice(
    priority: "low" | "medium" | "high" = "medium",
  ): Promise<this> {
    const gas = await this.client.getGasPrice();

    let tip: bigint;

    switch (priority) {
      case "low":
        tip = gas.priorityFeeLow;
        break;
      case "high":
        tip = gas.priorityFeeHigh;
        break;
      default:
        tip = gas.priorityFeeMedium;
    }

    this.params.maxPriorityFee = tip;
    this.params.maxFeePerGas = gas.getMaxFee();

    return this;
  }
  build(): CustomTransactionRequest {
    //Validate and return transaction request.

    if (!this.params.to) throw Error("Missing 'to'");
    if (!this.params.value) throw Error("Missing 'value'");

    const tx = new CustomTransactionRequest(this.params);
    return tx;
  }
  async buildAndSign(): Promise<string> {
    // Build, sign, and return ready-to-send transaction
    const tx = this.build();
    const signedTx = await this.wallet.signTransaction(tx);
    return signedTx;
  }

  async send(): Promise<string> {
    //Build, sign, send, return tx hash.
    const signedTx = await this.buildAndSign();
    const res = await this.client.sendTransaction(signedTx);
    return res;
  }
  async sendAndWait(
    timeout: number = 120,
  ): Promise<CustomTransactionReceipt | null> {
    // Build, sign, send, wait for confirmation.
    const tx_hash = await this.send();
    const res = await this.client.waitForReceipt(tx_hash, timeout);
    return res;
  }
}

File: chain/transaction_builder.py
import math
from chain.chain_client import ChainClient
from core.base_types import Address, TokenAmount, TransactionReceipt, TransactionRequest
from core.wallet import WalletManager

from eth_account.datastructures import SignedTransaction


class TransactionBuilder:
    """
    Fluent builder for transactions.

    Usage:
        tx = (TransactionBuilder(client, wallet)
            .to(recipient)
            .value(TokenAmount.from_human("0.1", 18))
            .data(calldata)
            .with_gas_estimate()
            .with_gas_price("high")
            .build())
    """

    params: dict

    def __init__(self, client: ChainClient, wallet: WalletManager):
        self.client: ChainClient = client
        self.wallet: WalletManager = wallet
        self.params = {}

    def to(self, address: Address) -> "TransactionBuilder":
        self.params["to"] = address
        return self

    def value(self, amount: TokenAmount) -> "TransactionBuilder":
        self.params["value"] = amount
        return self

    def data(self, calldata: bytes) -> "TransactionBuilder":
        self.params["data"] = calldata
        return self

    def nonce(self, nonce: int) -> "TransactionBuilder":
        """Explicit nonce (for replacement or batch)."""
        self.params["nonce"] = nonce
        return self

    def chain_id(self, number: int) -> "TransactionBuilder":
        """Explicit nonce (for replacement or batch)."""
        self.params["chain_id"] = number
        return self

    def gas_limit(self, limit: int) -> "TransactionBuilder":
        self.params["gas_limit"] = limit
        return self

    def with_gas_estimate(self, buffer: float = 1.2) -> "TransactionBuilder":
        """Estimate gas and set limit with buffer."""
        tx = TransactionRequest(
            to=self.params["to"],
            value=self.params.get("value", TokenAmount(0, 0, ",")),
            data=self.params["data"],
            nonce=self.params["nonce"],
        )
        estimated = self.client.estimate_gas(tx)
        self.params["gas_limit"] = math.ceil(estimated * buffer)
        return self

    def with_gas_price(self, priority: str = "medium") -> "TransactionBuilder":
        """Set gas price based on current network conditions."""
        gas = self.client.get_gas_price()
        match priority:
            case "low":
                tip = gas.priority_fee_low
            case "high":
                tip = gas.priority_fee_high
            case _:
                tip = gas.priority_fee_medium

        self.params["max_priority_fee"] = tip
        self.params["max_fee_per_gas"] = gas.get_max_fee()
        return self

    def build(self) -> TransactionRequest:
        """Validate and return transaction request."""
        try:
            tx = TransactionRequest(
                to=self.params["to"],
                value=self.params["value"],
                data=self.params["data"],
                nonce=self.params["nonce"],
                gas_limit=self.params["gas_limit"],
                max_fee_per_gas=self.params["max_fee_per_gas"],
                max_priority_fee=self.params["max_priority_fee"],
                chain_id=self.params.get("chain_id",1),
            )
        except:
            return None
        return tx

    def build_and_sign(self) -> SignedTransaction:
        """Build, sign, and return ready-to-send transaction."""
        tx = self.build()
        signed_tx = self.wallet.sign_transaction(tx)
        return signed_tx

    def send(self) -> str:
        """Build, sign, send, return tx hash."""
        stx = self.build_and_sign()
        tx_hash = self.client.send_transaction(stx)
        return tx_hash

    def send_and_wait(self, timeout: int = 120) -> TransactionReceipt:
        """Build, sign, send, wait for confirmation."""
        tx_hash = self.send()
        tr = self.client.wait_for_receipt(tx_hash=tx_hash, timeout=timeout)
        return tr
